#-*- mode: snakemake -*-
"""RSeQC: An RNA-seq Quality Control Package

http://rseqc.sourceforge.net/

"""
rule qc_bam_rseqc_read_distribution:
    input:
        bam = get_sorted_bam,
        bed = join(REF_DIR, 'genes', 'genes.bed12')
    singularity:
        'docker://quay.io/biocontainers/rseqc:3.0.0--py36hf8a1672_0'
    output:
        'logs/{sample}/{sample}.readdist.txt'
    shell:
        'read_distribution.py  -i {input.bam} -r {input.bed} > {output}'

rule qc_bam_rseqc_junction_annotation:
    input:
        bam = get_sorted_bam,
        bed = join(REF_DIR, 'genes', 'genes.bed12')
    singularity:
        'docker://quay.io/biocontainers/rseqc:3.0.0--py36hf8a1672_0'
    params:
        output = 'logs/{sample}/{sample}'
    shadow:
        'shallow'
    output:
        'logs/{sample}/{sample}.junction.xls'
    shell:
        'junction_annotation.py  -i {input.bam} -r {input.bed} -o {params.output}'
        
rule qc_bam_rseqc_junction_saturation:
    input:
        bam = get_sorted_bam,
        bed = join(REF_DIR, 'genes', 'genes.bed12')
    singularity:
        'docker://quay.io/biocontainers/rseqc:3.0.0--py36hf8a1672_0'
    params:
        output = 'logs/{sample}/{sample}'
    shadow:
        'shallow'
    output:
        'logs/{sample}/{sample}.junctionSaturation_plot.r'
    shell:
        'junction_saturation.py  -i {input.bam} -r {input.bed} -o {params.output}'

rule qc_bam_rseqc_inner_distance:
    input:
        bam = get_sorted_bam,
        bed = join(REF_DIR, 'genes', 'genes.bed12')
    singularity:
        'docker://quay.io/biocontainers/rseqc:3.0.0--py36hf8a1672_0'
    params:
        output = 'logs/{sample}/{sample}'
    output:
        'logs/{sample}/{sample}.inner_distance.txt'
    shell:
        'inner_distance.py -i {input.bam} -r {input.bed} -o {params.output}'

rule qc_bam_rseqc_tin:
    input:
        bam = get_sorted_bam,
        bed = join(REF_DIR, 'genes', 'genes.bed12')
    singularity:
        'docker://quay.io/biocontainers/rseqc:3.0.0--py36hf8a1672_0'
    output:
        'logs/{sample}/{sample}.tin.xls',
        'logs/{sample}/{sample}.summary.txt'
    shell:
        'tin.py -i {input.bam} -r {input.bed} && mv '       

       
        

