#-*- mode: snakemake -*-

"""
Snakemake rules for counting gene expression values from aligned data using featurecounts
"""

extra_conf_fn = srcdir('salmon.config')
if os.path.exists(extra_conf_fn):
    with open(extra_conf_fn) as fh:
        c  = yaml.load(fh) or {}
        update_config2(config, c)

FC_INTERIM = join(QUANT_INTERIM, 'featurecounts')
FC_PROCESSED = join(QUANT_PROCESSED, 'featurecounts')

## config params
read_orientation = config.get('read_orientation')
if read_orientation is None:
    LIB_TYPE = 'rf'
else:
    STRAND = '0' if strand == 'unstranded' else '-s 1' if strand == 'fr' else '-s 2'


rule featurecounts:
    input:
        bam = get_namesorted_bam
        gtf = join(REF_DIR, 'genes', 'genes.gtf')
    output:
        counts = join(FC_INTERIM, '{sample}', 'counts.txt')
    params:
        strand = STRAND,
        summary = join(FC_INTERIM, '{sample}', 'counts.txt.summary'),
        paired_end = ''
    log:
        'logs/{sample}/featurecounts.txt.summary'
    threads:
        16
    singularity:
        
    shell:
        'featureCounts '
        '-a {params.gtf} '
        '-o {output.counts} '
        '-T {threads} '
        '{params.strand} '
        '{params.paired_end}'
        '{input} '
        '&& '
        'mv {params.summary} {log}'

rule featurecounts_quant:
    input:
        counts = expand(rules.featurecounts.output.counts, sample=SAMPLES)
        gene_info = join(REF_DIR, 'genes', 'genes.csv')
    params:
        output = join(FC_PROCESSED, 'featurecounts', 'featurecounts'),
        script_dir = srcdir('scripts/')
    output:
        gene_counts = join(FC_PROCESSED, 'featurecounts', 'featurecounts.gene.quant'),
        tpm = join(FC_PROCESSED, 'featurecounts', 'featurecounts.gene.tpm'),
        gene_info = join(FC_PROCESSED, 'featurecounts', 'featurecounts.gene_info.tsv')
    shell:
        'Rscript {params.script_dir}/gene_quant.R '
        '{input.counts} '
        '-o {params.output} '
        '-t featurecounts '
        '--output-genelength '
        '--output-tpm '
        '--gene-info {input.gene_info} '
        '-v'
